### PR TITLE
nixos/luksroot: Check whether the device already exists

### PR DESCRIPTION
**Pull request by dasJ**
The new reuse behaviour is cool and really useful but it breaks one of
my use cases. When using kexec, I have a script which will unlock the
disks in my initrd. However, do_open_passphrase will fail if the disk is
already unlocked.

Btw, can this be backported?

###### Motivation for this change

It braks my old (18.03) kexec-and-autounlock behaviour.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Mic92** [comment]
@GrahamcOfBorg test boot.luksroot

**Comment by Mic92** [comment]
@GrahamcOfBorg test installer.luksroot

**Comment by dasJ** [comment]
Bump

File: nixos/modules/system/boot/luksroot.nix [hunk 1]
```diff
@@ -118,6 +118,11 @@ let
     do_open_passphrase() {
         local passphrase
 
+        # Return if already unlocked
+        if [ -b "/dev/mapper/${name}" ]; then
+          return
+        fi
```
Comment (Infinisil) [inline nixos/modules/system/boot/luksroot.nix]: Use `cryptsetup status ${name}` instead, it will exit with 0 if it exists (with 4 if it doesn't)

**Review by infinisil — CHANGES_REQUESTED** [review]

**Comment by dasJ** [comment]
@Infinisil Fixed and rebased

**Review by infinisil — APPROVED** [review]

**Comment by eadwu** [comment]
Broke my boot sequence, stuck on `waiting for device ...`.

**Comment by dasJ** [comment]
@eadwu Strange, to me it looks like it's waiting before the changed code is executed

**Comment by dasJ** [comment]
Yes, the order from my initrd:

```
wait_target …

do_open_passphrase() {
  # My code is here
}

open_normally() {
  do_open_passphrase
}

open_normally
```

**Comment by Mic92** [comment]
It would help to add `boot.trace` to the kernel `boot.kernelParams` to see what code it executes.

**Comment by Infinisil** [comment]
Ah darn, thanks for the quick report @eadwu. @Mic92 You gonna revert it on master as well?

**Pull request merged**